tests/test_filtering.py

```diff
@@ -0,0 +1,251 @@
+        result = (np.sin(R_ratio) - R_ratio * np.cos(R_ratio)) / (
+            2 * np.pi**2 * r_in**3
+        )
+        result[r_in == 0] = (
+            1 / 6 / np.pi**2 * (0.413566994 * R_filter) ** 3
+        )  # r->0 limit
+        return result
+    elif filter_flag == 2:
+        # output is Gaussian
+        const = (0.643 * R_filter) ** 2
+        exp_vol = (2 * np.pi * const) ** (3.0 / 2.0)
+        return np.exp(-((r_in) ** 2 / const / 2)) / exp_vol
+    elif filter_flag == 3:
+        # output is sphere with exponential damping
+        exp_vol = 4 / 3 * np.pi * R_filter**3
+        return (R_ratio < 1) * np.exp(-r_in / R_param) / exp_vol
+    elif filter_flag == 4:
+        # output is spherical shell
+        exp_vol = 4 / 3 * np.pi * (R_filter**3 - R_param**3)
+        return (R_ratio < 1) * (R_param <= r_in) / exp_vol
+
+
+# return binned mean & 1-2 sigma quantiles
+def get_binned_stats(x_arr, y_arr, bins, stats):
+    x_in = x_arr.flatten()
+    y_in = y_arr.flatten()
+    result = {}
+
+    statistic_dict = {
+        "pc1u": lambda x: np.percentile(x, 84),
+        "pc1l": lambda x: np.percentile(x, 16),
+        "pc2u": lambda x: np.percentile(x, 97.5),
+        "pc2l": lambda x: np.percentile(x, 2.5),
+        # used to mark percentiles in errorbar plots, since these cannot be negative
+        "err1u": lambda x: np.maximum(np.percentile(x, 84) - np.mean(x), 0),
+        "err2u": lambda x: np.maximum(np.percentile(x, 97.5) - np.mean(x), 0),
+        "err1l": lambda x: np.maximum(np.mean(x) - np.percentile(x, 16), 0),
+        "err2l": lambda x: np.maximum(np.mean(x) - np.percentile(x, 2.5), 0),
+    }
+
+    for stat in stats:
+        spstatkey = statistic_dict[stat] if stat in statistic_dict.keys() else stat
+        result[stat], _, _ = binstat(x_in, y_in, bins=bins, statistic=spstatkey)
+
+    return result
+
+
+@pytest.mark.parametrize("R", R_PARAM_LIST)
+@pytest.mark.parametrize("filter_flag", options_filter)
+def test_filters(filter_flag, R, plt):
+    opts = prd.get_all_options(redshift=10.0)
+
+    up = UserParams(opts["user_params"])
+    cp = CosmoParams(opts["cosmo_params"])
+    ap = AstroParams(opts["astro_params"])
+    fo = FlagOptions(opts["flag_options"])
+
+    # testing a single pixel source
+    input_box_centre = np.zeros((up.HII_DIM,) * 3, dtype="f4")
+    input_box_centre[up.HII_DIM // 2, up.HII_DIM // 2, up.HII_DIM // 2] = 1.0
+    output_box_centre = np.zeros((up.HII_DIM,) * 3, dtype="f8")
+    # use MFP=20 for the exp filter, use a 3 cell shell for the annular filter
+    if filter_flag == 3:
+        R_param = 20
+    elif filter_flag == 4:
+        R_param = max(R - 3 * (up.BOX_LEN / up.HII_DIM), 0)
+    else:
+        R_param = 0
+
+    lib.test_filter(
+        up(),
+        cp(),
+        ap(),
+        fo(),
+        ffi.cast("float *", input_box_centre.ctypes.data),
+        R,
+        R_param,
+        filter_flag,
+        ffi.cast("double *", output_box_centre.ctypes.data),
+    )
```

Comment:
can we write a low-level wrapper function for this in the main code?

Reply:
In fact, if instead we write the `filter_box` function as a low-level wrapper, we could perhaps write the `test_filter` function directly in python?

Reply:
Would the second option not require us to expose a bunch of fftw functions to the python wrapper?

Reply:
Not necessarily -- we could just use `np.fft` or even `pyfftw` to do the FFT stuff in the wrapper function

Reply:
Wouldn't we be testing different funcitons at that point?

Reply:
If you're okay with it, I would prefer to write wrappers for all the testing functions together in a separate PR

Reply:
That's fine with me